src/noble/element.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2021 Cloudflare, Inc. and contributors.
+// Copyright (c) 2021 Cloudflare, Inc.
+// Licensed under the BSD-3-Clause license found in the LICENSE file or
+// at https://opensource.org/licenses/BSD-3-Clause
+
+import { ProjPointType } from '@noble/curves/abstract/weierstrass'
+import { Elt } from '../groupTypes.js'
+import { ScalarNb } from './scalar.js'
+import { Point, PointConstructor } from './types.js'
+import type { GroupNb } from './group.js'
+import { compat, errDeserialization } from '../util.js'
+
+export class EltNb implements Elt {
+    Point: PointConstructor
+
+    private constructor(
+        public readonly g: GroupNb,
+        private readonly p: Point
+    ) {
+        this.Point = g.params.element.Point
+    }
+
+    static create(g: GroupNb): EltNb {
+        return new EltNb(g, g.params.element.Point.ZERO)
+    }
+
+    static gen(g: GroupNb): EltNb {
+        return new EltNb(g, g.params.element.Point.BASE)
+    }
+
+    isIdentity(): boolean {
+        return this.p.equals(this.Point.ZERO)
+    }
+
+    isEqual(a: EltNb): boolean {
+        compat(this, a)
+        return this.p.equals(a.p)
+    }
+
+    neg(): EltNb {
+        if (typeof this.p.negate === 'undefined') {
+            // https://github.com/paulmillr/noble-curves/issues/84
+            throw new Error(`Point doesn't implement negate: ${this.Point}`)
+        } else {
+            return new EltNb(this.g, this.p.negate())
+        }
+    }
+
+    add(a: EltNb): EltNb {
+        compat(this, a)
+        return new EltNb(this.g, this.p.add(a.p))
+    }
+
+    mul(s: ScalarNb): EltNb {
+        compat(this, s)
+        return new EltNb(this.g, this.p.multiply(s.k))
+    }
+
+    mul2(k1: ScalarNb, a: EltNb, k2: ScalarNb): EltNb {
+        compat(this, k1)
+        compat(this, k2)
+        compat(this, a)
+        if (this.p.multiplyAndAddUnsafe) {
+            const zero = this.Point.ZERO
+            const el = this.p.multiplyAndAddUnsafe(a.p, k1.k, k2.k) ?? zero
+            return new EltNb(this.g, el)
+        } else {
+            // Manually perform k1 * this.p + k2 * a.p
+            const term1 = this.p.multiply(k1.k)
+            const term2 = a.p.multiply(k2.k)
+            const sum = term1.add(term2)
+            return new EltNb(this.g, sum)
+        }
+    }
+
+    serialize(compressed = true): Uint8Array {
+        if (this.isIdentity() && !this.g.params.isEdwards) {
```

Comment:
It's probably better to first ask whether it is not an Edwards point, if that's the case, check for identity.